sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusReceiverClient.java

```diff
@@ -719,28 +721,30 @@ private void queueWork(int maximumMessageCount, Duration maxWaitTime,
 
         final long id = idGenerator.getAndIncrement();
         final SynchronousReceiveWork work = new SynchronousReceiveWork(id, maximumMessageCount, maxWaitTime, emitter);
-        final SynchronousMessageSubscriber messageSubscriber = synchronousMessageSubscriber.get();
+        SynchronousMessageSubscriber messageSubscriber = synchronousMessageSubscriber.get();
 
         if (messageSubscriber != null) {
             messageSubscriber.queueWork(work);
             return;
         }
 
-        final SynchronousMessageSubscriber newSubscriber = new SynchronousMessageSubscriber(asyncClient,
-            work,
-            isPrefetchDisabled,
-            operationTimeout);
-
         // NOTE: We asynchronously send the credit to the service as soon as receiveMessage() API is called (for first
         // time).
         // This means that there may be messages internally buffered before users start iterating the IterableStream.
         // If users do not iterate through the stream and their lock duration expires, it is possible that the
         // Service Bus message's delivery count will be incremented.
-        if (synchronousMessageSubscriber.compareAndSet(null, newSubscriber)) {
-            asyncClient.receiveMessagesNoBackPressure().subscribeWith(newSubscriber);
-        } else {
-            newSubscriber.dispose();
-            synchronousMessageSubscriber.get().queueWork(work);
+        synchronized (newSubscriberLock) {
+            messageSubscriber = synchronousMessageSubscriber.get();
+            if (messageSubscriber == null) {
+                final SynchronousMessageSubscriber newSubscriber = new SynchronousMessageSubscriber(asyncClient,
+                    work,
+                    isPrefetchDisabled,
+                    operationTimeout);
+                synchronousMessageSubscriber.compareAndSet(null, newSubscriber);
+                asyncClient.receiveMessagesNoBackPressure().subscribeWith(newSubscriber);
+            } else {
+                messageSubscriber.queueWork(work);
+            }
```

Comment:
I understand this `subscribeWith` and `queueWork` ops happen only once. But I was wondering, instead of `synchronized`, if we could use an approach close to what we originally had?

i.e. replacing `newSubscriberLock` with

```java
private final AtomicBoolean syncSubscribed = new AtomicBoolean(false);
```

then make use of API in AtomicReference to ensure we instantiate subscriber only once + AtomicBoolean to ensure we subscribe only once.

```java
SynchronousMessageSubscriber syncSubscriber = synchronousMessageSubscriber.get();
if (syncSubscriber == null) {
    syncSubscriber.queueWork(work);
    return;
}

syncSubscriber = synchronousMessageSubscriber.updateAndGet(subscriber -> {
    // Ensuring we create SynchronousMessageSubscriber only once.
    if (subscriber == null) {
        return new SynchronousMessageSubscriber(asyncClient,
            work,
            isPrefetchDisabled,
            operationTimeout);
    } else {
        return subscriber;
    }
});

if (!syncSubscribed.getAndSet(true)) {
    // The 'subscribeWith' has side effects hence must not be called from
    // the above updateFunction of AtomicReference::updateAndGet.
    asyncClient.receiveMessagesNoBackPressure().subscribeWith(syncSubscriber);
} else {
    syncSubscriber.queueWork(work);
}
```

Reply:
the main idea is to avoid holding a strong lock when calling into these operations.

Reply:
Thanks for the detailed suggestion. It looks great!
 I'll try to use this way to avoid using lock.